### PR TITLE
[Torch Ranker Agent] Caching candidate encoding

### PR DESCRIPTION
**Pull request by emilydinan**
Change the `score_candidates` API to possibly have `cand_encs` passed in. This occurs in the case when we want to cache the candidate encodings (possibly when using fixed or vocab candidates). The eval_step in TRA will pass these into score_candidates. It is then up to the model to implement `score_candidates`. See the BiEncoderRanker agent or TransformerRanker agent to see how this is implemented.

Also included, bug fix caused by conflicts in #1404.

cc @samhumeau

### CODE REVIEW

File: parlai/agents/bert_ranker/bi_encoder_ranker.py [hunk 1]
```diff
@@ -69,9 +69,11 @@ def set_vocab_candidates(self, shared):
         self.vocab_candidates will contain a [num_cands] list of strings
         self.vocab_candidate_vecs will contain a [num_cands, 1] LongTensor
         """
+        self.opt['encode_candidate_vecs'] = True
         if shared:
             self.vocab_candidates = shared['vocab_candidates']
             self.vocab_candidate_vecs = shared['vocab_candidate_vecs']
+            self.vocab_candidate_encs = shared['vocab_candidate_encs']
```
Comment (stephenroller) [inline parlai/agents/bert_ranker/bi_encoder_ranker.py]: dupe

File: parlai/agents/bert_ranker/bi_encoder_ranker.py [hunk 2]
```diff
@@ -135,23 +137,16 @@ def _set_text_vec(self, *args, **kwargs):
...
-        if (hasattr(self, 'fixed_candidate_encs') and
-                self.fixed_candidate_encs is not None):
-            return embedding_ctxt.mm(self.fixed_candidate_encs.t())
-
-        # evaluating vocab candidates:
-        if (hasattr(self, 'vocab_candidate_encs') and
-                self.vocab_candidate_encs is not None):
-            return embedding_ctxt.mm(self.vocab_candidate_encs.t())
+        if cand_encs is not None:
+            return embedding_ctxt.mm(cand_encs.t())
```
Comment (stephenroller) [inline parlai/agents/bert_ranker/bi_encoder_ranker.py]: this reads much better

File: parlai/agents/transformer/modules.py [hunk 3]
```diff
@@ -144,6 +144,9 @@ def encode_cand(self, words):
 
     def encode_context_memory(self, context_w, memories_w):
         # [batch, d]
+        if context_w is None:
```
Comment (stephenroller) [inline parlai/agents/transformer/modules.py]: Why did this change here?

Comment (stephenroller) [reply parlai/agents/transformer/modules.py]: oh nvm, I see below. So Transformer wasn't even using the caching before?

Comment (stephenroller) [reply parlai/agents/transformer/modules.py]: Actually, I'm still confused. Nothing about context memory should ever be cached right?

File: parlai/core/torch_ranker_agent.py [hunk 4]
```diff
@@ -119,8 +112,16 @@ def __init__(self, opt, shared=None):
                 broadcast_buffers=False,
             )
 
-    def score_candidates(self, batch, cand_vecs):
-        """Given a batch and candidate set, return scores (for ranking)"""
+    def score_candidates(self, batch, cand_vecs, cand_encs=None):
+        """Given a batch and candidate set, return scores (for ranking).
+
+        :param batch:       a Batch object (defined in torch_agent.py)
```
Comment (stephenroller) [inline parlai/core/torch_ranker_agent.py]: Mind if we use the new docstring format?

https://github.com/facebookresearch/ParlAI/blob/baa839db2120c837562620386cbe1167bd0e9109/parlai/core/torch_agent.py#L650-L659

File: parlai/agents/bert_ranker/bi_encoder_ranker.py (hunk 1 above)
Comment (emilydinan) [reply parlai/agents/bert_ranker/bi_encoder_ranker.py]: sorry i don't understand this comment?

File: parlai/agents/transformer/modules.py (hunk 3 above)
Comment (emilydinan) [reply parlai/agents/transformer/modules.py]: correct - transformer was not caching before. 

this is just saying if context is None, we want to return None. (useful for when we want to encode just the RHS)

File: parlai/agents/bert_ranker/bi_encoder_ranker.py (hunk 1 above)
Comment (stephenroller) [reply parlai/agents/bert_ranker/bi_encoder_ranker.py]: Oh I thought it was a duplicate of the line above but I misread. Please disregard

File: parlai/agents/transformer/modules.py (hunk 3 above)
Comment (stephenroller) [reply parlai/agents/transformer/modules.py]: So I’m still a tad bit confused why then this guard ends up in the encode_context_memory method

Comment (stephenroller) [reply parlai/agents/transformer/modules.py]: Anyway maybe you can just add a short comment here saying why context_w might be None and then it lgtm.

Comment (samhumeau) [reply parlai/agents/transformer/modules.py]: "this is just saying if context is None, we want to return None. "
I am confused. If context is None, shouldn't we return the embedding of the candidates only?

Comment (samhumeau) [reply parlai/agents/transformer/modules.py]: Ah, nevermind, wrong function

**Pull request merged**